### PR TITLE
feat(obs): add indexer_message_latency_seconds by stage

### LINKED ISSUE #40: feat(obs): add end-to-end message latency metric to es-indexer consumer
## Background

现有 consumer/producer metrics 只覆盖**处理耗时**（`indexer_io_op_duration_seconds`、`searchetl_producer_tick_duration_seconds` 等），无法回答业务最关心的问题：**一条消息从产生到可被检索，到底要多久**。当前缺少**消息延迟**（end-to-end，墙钟）这一维度的可观测性。

## Proposal

在 consumer 侧新增单个 Histogram，用 `stage` label 区分两种延迟，避免指标爆炸：

- `indexer_message_latency_seconds{stage="e2e"}` = `now - created_at`（DB 落库 → ES 可见）
- `indexer_message_latency_seconds{stage="send_to_index"}` = `now - msg_timestamp`（消息发送 → ES 可见）

埋点位置：consumer 写 ES 成功（`dispOK`）那一刻 —— 此时 `CreatedAt` / `MsgTimestamp`（均为秒级 epoch）与处理时刻 `now` 同时在手，无需改 Kafka 协议、无需新增表。

## Design notes

- 单一指标名 + `stage` label，控制 label 基数。
- 延迟跨度比处理耗时大，使用独立 bucket（`[0.5, 1, 2.5, 5, 10, 30, 60, 120, 300, 600, 1800]`，0.5s ~ 30min），与处理耗时 bucket（5ms~5s）分开。
- 负值（时钟漂移导致 `now < ts`）clamp 到 0；`CreatedAt` / `MsgTimestamp` 为 0 时跳过对应 stage 的观测，避免把「缺时间戳」当成超大延迟污染分位。
- 时钟依赖：源时间戳来自 DB/业务时钟，consumer 用本机 `now`，依赖 NTP 同步。
- 告警阈值需结合 `PRODUCER_LAG_SECONDS`（producer 稳定性 lag 窗口，配置项，线上不一定是默认值）+ 业务 SLA 设定，**不应绑死任何具体 lag 值**。

## Why not kafka-exporter consumer lag

kafka-exporter 的 consumer lag 是 **offset 条数差**，不是时间维度，回答不了「用户多久能搜到」。本指标是时间维度，直接对应业务 SLA；两者可并存。

## Scope

- 仅 consumer 侧埋点，最小 diff，遵循现有 `ObserveIO` / `MarkIOError` 的风格，不新增文件、不引入抽象层。
- 已在本地分支实现并通过验证（`go build ./...` / `go vet ./internal/consumer/...` / `go test ./internal/consumer/...` 全绿），可提 PR。

### PR DESCRIPTION
**Pull request by mou-ren**
Closes #40

## What

在 es-indexer consumer 侧新增单个 Histogram，度量消息到 ES 可见的延迟，用 `stage` label 区分两种：

- `indexer_message_latency_seconds{stage="e2e"}` = `now - created_at`（DB 落库 → ES 可见）
- `indexer_message_latency_seconds{stage="send_to_index"}` = `now - msg_timestamp`（消息发送 → ES 可见）

## Where

埋点在 `resolvePass` 的 `dispOK` 分支（ES bulk 成功那一刻）——此时 `CreatedAt` / `MsgTimestamp`（秒级 epoch）与处理时刻 `now` 同时在手，无需改 Kafka 协议、无需新表。

## Design notes

- 单一指标名 + `stage` label，控制基数。
- 独立延迟 bucket（`[0.5, 1, 2.5, 5, 10, 30, 60, 120, 300, 600, 1800]`，0.5s~30min），与处理耗时 bucket（5ms~5s）分开。
- 负值（时钟漂移 `now < ts`）clamp 到 0；源时间戳为 0 时跳过对应 stage，避免把缺时间戳当成超大延迟。
- 风格对齐现有 `ObserveIO` / `MarkIOError`，无新增文件、无抽象层，最小 diff。

## Test

- `go build ./...` ✅
- `go vet ./internal/consumer/...` ✅
- `go test ./internal/consumer/...` ✅
- 全量 `go test ./...` ✅

### CODE REVIEW

**Review by OctoBoooot — COMMENTED** [review]
## Review: feat(obs): add indexer_message_latency_seconds by stage (#41)

**Verdict:** Comment (code is clean; only a non-code CI gate blocks APPROVE)

Small, correct follow-up to #39 (2 files, +30/-1). The code has no findings. The only red check is `check-sprint` — the maintainer governance gate (linked issue not on a sprint), not a code issue — so per the CI-red precondition this is COMMENT, not APPROVE. All actual code checks pass.

Risk tier: high (consumer path), but this is purely additive instrumentation hooked only on the successful-index disposition; no correctness path touched.

### What I verified

- **Timestamp units are correct** — the load-bearing risk for a latency metric. Both `parsed[idx].CreatedAt` and `MsgTimestamp` are documented epoch **seconds** in the contract (`octo-lib/contract/searchmsg/searchmsg.go:94-97`: "纪元秒"), so `time.Unix(ts, 0)` is the right conversion. No ms/s mismatch → no 1000× latency inflation.
- **Clock-skew guarded** — `ObserveMessageLatency` clamps negative durations to 0 (`now < ts`), so a skewed clock can't emit a nonsense negative sample.
- **Unset-timestamp guarded** — the hooks only fire when `CreatedAt > 0` / `MsgTimestamp > 0`, so a message missing a timestamp doesn't record a bogus `now - epoch(0)` (~56-year) latency. Good — this is the trap with `time.Unix` on a zero value.
- **Label cardinality bounded** — `stage` is a 2-value enum (`e2e` / `send_to_index`); no unbounded series.
- **Buckets fit the domain** — separate `msgLatencyBuckets` (0.5s–1800s) for end-to-end latency vs the IO-op 5ms–5s set; registered correctly on the same private registry, nil-safe via `if p.metrics != nil`, hooked only on `dispOK`.

### Note (non-blocking)

- No unit test for `ObserveMessageLatency` (the clamp + the stage labeling). Trivial to add given the existing `metrics_test.go` harness — a 1-case test asserting the negative-clamp and the two stages would lock in the clock-skew guard. Not a blocker for a 30-line additive metric, but cheap insurance.

### Praise

- The two guards — negative-clamp and `>0` timestamp checks — are exactly the two ways `time.Unix`-based latency metrics go wrong (clock skew → negative; zero-value field → epoch-distance), and both are handled up front rather than discovered via a garbage dashboard later. For a small obs PR that's precisely the right paranoia.

**Review by mochashanyao — APPROVED** [review]
[Octo-Q · automated review]

Verdict: Approve — no blocking findings; notes below (data-flow traced).

---

# octo-search-indexer PR#41 Review Report
**Reviewer**: Octo-Q (automated review)
**Head SHA:** `b75ed4904bc3fc5e4f9eb84f4a6dc0dc27ab7598`
**Repo:** Mininglamp-OSS/octo-search-indexer
**PR:** feat(obs): add indexer_message_latency_seconds by stage

## 1. 验证结论

- ✅ **Metric definition correct** — `indexer_message_latency_seconds` histogram registered in private registry with `stage` label, namespace `indexer`, buckets `[0.5, 1, 2.5, 5, 10, 30, 60, 120, 300, 600, 1800]` (`metrics.go:89-97`)
- ✅ **Instrumentation placement correct** — observed only in `dispOK` case (`consumer.go:261-270`), after ES bulk success confirmation. No double-counting: `dispositions[idx]` transitions from `dispTransient` to `dispOK` exactly once per message across retry loops
- ✅ **Negative duration clamp** — `ObserveMessageLatency` at `metrics.go:148-150` clamps `d < 0` to 0 before `Observe()`. Handles clock skew correctly
- ✅ **Zero-timestamp guard** — `CreatedAt > 0` and `MsgTimestamp > 0` checks at `consumer.go:263,266` correctly skip messages missing timestamps (JSON zero-value = absent)
- ✅ **Nil-guard pattern** — `p.metrics != nil` at `consumer.go:261` matches existing convention (`ObserveIO`, `MarkIOError` etc.)
- ✅ **No control-flow change** — purely additive observability code; existing bulk/DLQ/commit paths unchanged

## 2. 发现的问题

### P2-1: Test coverage gap for new metric path (confidence: 9/10)

**Diff-scope:** new (this PR introduces the metric but no test exercises it)

The `validMsg()` test helper (`consumer_test.go:159-171`) constructs `searchmsg.Message` with `CreatedAt: 0` and `MsgTimestamp: 0` (Go zero values). The new `> 0` guards at `consumer.go:263,266` will skip both `ObserveMessageLatency` calls for all existing test fixtures.

Result: `TestProcessBatch_DispositionMetrics` and all other integration tests pass without ever hitting the new metric code. The `ObserveMessageLatency` method and its integration into `resolvePass` are untested.

**Suggested fix:** Add a test (e.g., `TestProcessBatch_MessageLatencyMetric`) that constructs messages with non-zero `CreatedAt` and `MsgTimestamp`, runs `processBatch`, and asserts `indexer_message_latency_seconds_count{stage="e2e"}` and `{stage="send_to_index"}` appear in the registry output. Pattern mirrors existing `TestProcessBatch_DispositionMetrics`.

### P2-2: Bucket coverage for consumer lag recovery (confidence: 7/10)

**Diff-scope:** new (design choice in this PR)

Max bucket is 1800s (30 min). During Kafka consumer lag recovery (indexer was down, catches up on backlog), messages can be hours old. Those will land in `+Inf` bucket, losing percentile granularity for lag-recovery episodes. Not a correctness issue (Prometheus histograms always have `+Inf`), but p99 latency during catch-up will be less informative.

This is an operational awareness note, not a blocker. The current bucket range is well-suited for steady-state real-time indexing.

## 3. 建议

1. Add a dedicated test for the latency metric path (see P2-1 above). Low effort, high value for regression protection.
2. Consider documenting the bucket design rationale in a code comment or PR description for future operators who may wonder about the 30min cap during lag recovery.

## 4. 额外发现

None. The diff is minimal (30 additions, 1 deletion across 2 files), follows established patterns, and introduces no new abstractions or side effects.

## 5. 数据流回溯

| Consumed data | Upstream source | Path to consumption | Verified |
|---|---|---|---|
| `parsed[idx].CreatedAt` | `searchmsg.Message.CreatedAt` (JSON `created_at`, int64 epoch seconds) from producer (ETL/CDC) via `message.created_at` DB column | `decodeAndValidate` → `json.Unmarshal` → `parsed[i].CreatedAt` → `> 0` guard → `time.Unix(sec, 0)` → `now.Sub()` | ✅ Data flows correctly; zero-value guard prevents phantom observations |
| `parsed[idx].MsgTimestamp` | `searchmsg.Message.MsgTimestamp` (JSON `msg_timestamp`, int64 epoch seconds) from producer via `message.timestamp` DB column | `decodeAndValidate` → `json.Unmarshal` → `parsed[i].MsgTimestamp` → `> 0` guard → `time.Unix(sec, 0)` → `now.Sub()` | ✅ Data flows correctly; same guard pattern |
| `now` (observation time) | `time.Now()` at `consumer.go:262` | Called per-item in `bulkIdx` loop, after ES bulk result processed | ✅ Correctly represents ES-visibility moment |
| `msgLatency` histogram | `prometheus.NewHistogramVec` at `metrics.go:89-97` | Registered via `reg.MustRegister` at `metrics.go:101` in private registry | ✅ Will appear as `indexer_message_latency_seconds` |

## 6. 盲点 checklist (R5)

- **C1 双路径 parity:** N/A — pure observability addition, no symmetric create/delete/subscribe/unsubscribe paths touched
- **C2 control-flow ordering / 嵌套复用:** Clear — `ObserveMessageLatency` is called once per dispOK transition, after `markDisposition`. No interaction with the retry loop (dispositions are checked `!= dispTransient` at loop entry, so already-resolved items are skipped on retry). No nesting or double-observation risk
- **C3 授权边界 ≠ 能力边界:** N/A — no auth/permission/jail changes
- **C4 授权生命周期 / 容器-成员状态级联:** N/A — no auth changes
- **C5 build/note 通过 ≠ 运行期路径正确:** N/A — no build artifacts, extensions, or relative path changes
- **C6 治理/策略/安全文档自洽性:** N/A — no governance/policy documentation changes

## 7. 跨轮 blocker 复检

N/A — first review of this PR.

---

**[Octo-Q] verdict: APPROVE**

No P0/P1 findings. Two P2 notes: test coverage gap for the new metric path (should-fix before merge) and bucket upper-bound awareness for lag-recovery scenarios (informational). The change is clean, minimal, follows established patterns, and introduces no behavioral risk to existing paths.

**Review by yujiawei — APPROVED** [review]
# Code Review — PR #41 (octo-search-indexer)

Reviewed at head SHA `b75ed49`. Independent review against `origin/main` (merge-base `a29e074`). Diff is small and contained: +30/-1 across `internal/consumer/metrics.go` and `internal/consumer/consumer.go`. `go build ./...`, `go vet ./internal/consumer/...`, and `go test ./internal/consumer/...` all pass locally.

## 1. Spec compliance

**Spec: ✅**

The change implements exactly what the linked issue asks for, nothing more:

- Single `HistogramVec` `indexer_message_latency_seconds` with a `stage` label, two stages — `e2e` (`now - created_at`) and `send_to_index` (`now - msg_timestamp`). `internal/consumer/metrics.go:93-98`, `internal/consumer/consumer.go:262-269`.
- Dedicated coarse latency buckets `[0.5 … 1800]`, kept separate from the IO-op buckets. `internal/consumer/metrics.go:16-18`.
- Negative durations (clock skew) clamped to 0. `internal/consumer/metrics.go:146-151`.
- Source timestamp of 0 skips that stage, so a missing timestamp is not recorded as a huge latency. `internal/consumer/consumer.go:264,267`.
- Style matches existing `ObserveIO` / `MarkIOError`; private registry registration updated. `internal/consumer/metrics.go:101-103`.

No missing scope, no over-build (no extra flags/fields/labels), no deviation from the agreed approach.

**Cross-file semantics verified:** `time.Unix(field, 0)` treats both fields as epoch seconds. That is correct — `CreatedAt` / `MsgTimestamp` on `searchmsg.Message` are documented as epoch seconds, and the producer populates them from `UNIX_TIMESTAMP(created_at)` and the row `timestamp` column (`internal/producer/source.go:86-87`, `internal/producer/extract.go:57-58,94-95`). No unit mismatch.

**Observation point verified:** the observe call sits in the `case ok` branch of the per-item bulk-result loop, keyed by `bulkIdx`/`parsed[idx]`, so indices align with the bulk request and each message is observed once per successful index within a batch (`internal/consumer/consumer.go:250-270`). Once an item is `dispOK` it is excluded from `toBulk` on subsequent passes, so no intra-batch double-count.

## 2. Code quality

**Quality: Approved**

No P0/P1. The points below are all advisory (P2) — none should block merge.

- **P2 — at-least-once re-observation across redelivery.** The sample is recorded on bulk-item success, before the Kafka offset commit (`internal/consumer/consumer.go:262` vs the commit at `internal/consumer/consumer.go:~324`). If commit fails or the process crashes before commit, the batch is redelivered and the same logical message is observed again. For a latency histogram this is acceptable (minor count inflation, not correctness), but worth noting.

- **P2 — help text vs. actual measurement point.** `Help: "Message latency to ES visibility…"` (`internal/consumer/metrics.go:95`) but the sample is taken at ES bulk-item success, which precedes refresh/search visibility. Consider rewording to "to ES bulk-index success" to avoid implying the metric includes refresh lag.

- **P2 — future-dated source timestamps masked.** Clamping `d < 0` to 0 (`internal/consumer/metrics.go:147`) means a source clock ahead of the indexer silently produces zero-latency samples, hiding skew rather than surfacing it. Acceptable as a deliberate trade-off; a separate skew counter would make the skew observable, but that is out of scope here.

- **P2 — no test coverage for the new path.** `internal/consumer/metrics_test.go` does not assert that `indexer_message_latency_seconds` is registered, that it emits only `stage="e2e"` / `stage="send_to_index"`, that a zero source timestamp skips the stage, or that a negative duration clamps to 0. The existing `validMsg` test helper leaves `CreatedAt`/`MsgTimestamp` at 0, so the new branch is never exercised by the suite. A small direct test on `ObserveMessageLatency` plus a `dumpMetrics` assertion would lock in the clamp/skip behavior.

- **P2 — minor: hoist `time.Now()`.** `now := time.Now()` is read per item inside the loop (`internal/consumer/consumer.go:263`); it could be hoisted above the loop for one clock read and a consistent batch reference. Negligible impact.

- **P2 — `stage` label is caller-disciplined only.** `ObserveMessageLatency(stage string, …)` accepts any string, so cardinality safety relies on call sites. Both current callers pass string literals, so this is fine today; just keep that invariant if more stages are added.

## 3. Overall verdict

**APPROVE.** Spec ✅ and no blocking quality findings. The two suggestions most worth picking up before or shortly after merge are the test coverage for the clamp/skip behavior and the help-text wording, but neither blocks merge.

## 4. Coverage note

Reviewed: the full diff, the surrounding `resolvePass` flow and bulk-result index alignment, the `searchmsg.Message` field contract, and the producer SQL/extract path that populates the timestamps. Not separately exercised: live Prometheus scrape output and runtime clock-skew behavior in production (reasoned about statically only).

**Review by Jerry-Xin — APPROVED** [review]
Clean, well-guarded additive latency metric. No code blockers; the only red check is the non-code maintainer sprint gate.

**✅ Verified clean**
- **Cardinality**: the new `message_latency_seconds` histogram is labeled only by `stage`, which takes two literal constant values (`"e2e"`, `"send_to_index"`) at the two `ObserveMessageLatency` call sites (consumer.go:265,268). Bounded — no cardinality blowup.
- **Unit**: `ObserveMessageLatency` does `Observe(d.Seconds())` against a `time.Duration`, correct for the `_seconds` suffix. The durations are `now.Sub(time.Unix(parsed[idx].CreatedAt, 0))` / `MsgTimestamp` — and both fields are documented as epoch **seconds** (`int64`) in the contract (octo-lib `contract/searchmsg/searchmsg.go:94-97: 纪元秒`), so `time.Unix(ts, 0)` is the right conversion (no ms/s 1000× inflation).
- **`time.Unix` traps both guarded**: `if CreatedAt > 0` / `if MsgTimestamp > 0` prevent emitting `now - epoch(0)` (~56-year bogus latency) on an unset timestamp; and `ObserveMessageLatency` clamps negative durations to 0 for clock skew (`now < ts`).
- **Registration/placement**: `msgLatency` added to the single `reg.MustRegister(...)` in `NewMetrics` (once per binary, private registry — no double-register), nil-guarded (`if p.metrics != nil`), hooked only on the `dispOK` (successful-index) path.
- **Buckets**: a coarser 0.5s–1800s set (separate `msgLatencyBuckets` var) appropriate for end-to-end message latency, vs the IO 5ms–5s set — reasonable.
- **Naming**: `indexer_message_latency_seconds` follows Prometheus conventions (base unit seconds, `_seconds` suffix).

**💬 Non-blocking**
- 🟡 No test for the new metric path. `ObserveMessageLatency` (the negative-clamp + stage labeling) and the two hooks are untested; existing consumer tests don't exercise them. For a 30-line additive metric on an already-tested disposition path this is a coverage gap, not a correctness/data/security defect — worth a small unit test (assert the clamp and the two stage series), but it doesn't gate merge.
- 🔵 The 1800s (30min) bucket max loses granularity past 30min of lag-recovery latency; acceptable for e2e latency, fine as-is.

**Gate note**
- All code checks are green (Lint/Build/Test/Vet/CodeQL pass). The only failing check is `check-sprint`, a non-code maintainer governance gate (the linked issue needs a Sprint set on the Octo Board) — not a code change. It keeps `mergeStateStatus=UNSTABLE`, so a Maintainer needs to attach the issue to the sprint before this goes mergeable; the code itself is ready.

**Comment by lml2468** [comment]
# QA Review - fail

**Verdict**: `fail`  
**AC source**: linked issue #40  
**CI status**: Test/Build/Vet/Lint green; overall PR checks unstable because `check-sprint` failed on linked issue sprint metadata  
**Test diff**: +0 files, ~0 lines (no exemption)

## One-line conclusion

The code implements the requested metric path, but the PR adds no tests for the new observable behavior, so QA must fail closed.

## AC coverage matrix

| AC # | Description | Code path | Test | Assertion strength | Status |
|---|---|---|---|---|---|
| AC1 | Add one `indexer_message_latency_seconds` HistogramVec with bounded `stage` label values `e2e` and `send_to_index`. | `internal/consumer/metrics.go:92` | Missing | n/a | **MISSING TEST** |
| AC2 | Observe `e2e = now - created_at` and `send_to_index = now - msg_timestamp` only after successful ES bulk item handling (`dispOK`). | `internal/consumer/consumer.go:259` | Missing; existing `validMsg` fixtures leave both timestamps at zero | n/a | **MISSING TEST** |
| AC3 | Use dedicated latency buckets `[0.5, 1, 2.5, 5, 10, 30, 60, 120, 300, 600, 1800]`. | `internal/consumer/metrics.go:16` | Missing | n/a | **MISSING TEST** |
| AC4 | Clamp negative durations to 0 and skip a stage when its source timestamp is 0. | `internal/consumer/consumer.go:264`, `internal/consumer/metrics.go:147` | Missing | n/a | **MISSING TEST** |
| AC5 | Keep the change consumer-side and aligned with existing metrics style, without protocol/table changes. | `internal/consumer/consumer.go`, `internal/consumer/metrics.go` | Existing consumer tests still pass in CI, but do not cover the new branch | Weak for this PR | Partial |

## Boundary / edge cases

| Path | Boundary category | Tested? |
|---|---|---|
| `ObserveMessageLatency` | negative duration / clock skew clamps to 0 | **No** |
| `resolvePass` metric hook | zero `CreatedAt` / `MsgTimestamp` skips metric emission | **No** |
| `resolvePass` metric hook | both non-zero timestamps emit exactly the two expected stage series on successful bulk item | **No** |
| `msgLatency` histogram | bucket registration and bounded `stage` label behavior | **No** |

## Regression risk

- **Touched module**: `internal/consumer` observability path.
- **Recent churn**: changed in multiple recent commits, including the prior Prometheus metrics addition and this PR.
- **Public contract change**: new Prometheus metric series, which is an operational contract for dashboards/alerts.
- **Risk rating**: Medium. The runtime path is additive, but the new exported metric behavior lacks regression coverage.

## Observability

- [x] Metric added for the requested business latency dimension.
- [x] Label cardinality is bounded by two call-site literals today.
- [ ] Alert rule not added; acceptable because issue #40 explicitly says thresholds should be set from `PRODUCER_LAG_SECONDS` + SLA rather than hardcoded here.
- [ ] Skew/zero timestamp behavior is not separately observable; acceptable as a design tradeoff, but it needs tests because it is part of the requested behavior.

## Test execution

- **Existence**: No added or modified tests in this PR.
- **CI**: GitHub CI Test/Build/Vet/Lint jobs passed; `check-sprint` failed because linked issue #40 has no Sprint set on the Octo Board.
- **Coverage change**: No coverage signal available.
- **Flaky tests**: None identified.

## Blockers

1. Add regression tests for the new metric behavior. At minimum, assert that `indexer_message_latency_seconds_count{stage="e2e"}` and `{stage="send_to_index"}` are emitted for a successful bulk item with non-zero `CreatedAt` and `MsgTimestamp`.
2. Add direct coverage for the boundary behavior: negative durations clamp to 0, and zero source timestamps skip the corresponding stage instead of recording an epoch-distance latency.

<sub>Labels: `review:done:qa:fail` · Generated by qa-engineer · Not a merge approval.</sub>

**Comment by lml2468** [comment]
## 🔒 Security Review — CLEARED

**Verdict: `CLEARED`** — no new exploitable risk introduced.

### Scope
Two-file, additive observability change: a new `indexer_message_latency_seconds` Histogram (`stage` label) plus its emission point in `resolvePass`'s `dispOK` branch.

### Findings
- **Credential / secret scan**: clean — no secrets, tokens, or keys in the diff.
- **DoS / metric cardinality**: the `stage` label takes only two hardcoded literal values (`"e2e"`, `"send_to_index"`); no untrusted input flows into a metric label, so no Prometheus label-cardinality explosion. Buckets are bounded; timestamps are observed as histogram *values*, not labels.
- **STRIDE**: no new auth/identity surface, no DB/shell/template sinks, no deserialization. The change reads two numeric fields already in memory and records a duration.
- **Info disclosure**: exported series carry durations only — no sensitive fields in labels or values; Help text is benign.
- **Robustness**: `p.metrics != nil` guard prevents nil deref; negative-duration clamp (clock skew) and zero-timestamp skip are sound and avoid nonsensical samples.

No injection, broken access control, privilege, or logging-failure concerns apply to this surface.

> Note: the separate QA `fail` (missing tests for the new metric path) is a test-coverage / quality gate, not a security finding, and does not affect this security verdict.

File: internal/consumer/metrics.go [hunk 1]
```diff
@@ -130,3 +141,12 @@ func (m *Metrics) ObserveIO(op string, d time.Duration) {
 func (m *Metrics) MarkIOError(op string) {
 	m.ioOpErrors.WithLabelValues(op).Inc()
 }
+
+// ObserveMessageLatency records one message latency sample for stage (e2e or
+// send_to_index). Negative durations (clock skew where now < ts) clamp to 0.
+func (m *Metrics) ObserveMessageLatency(stage string, d time.Duration) {
```
Comment (lml2468) [inline internal/consumer/metrics.go:147]: Should Fix: please add focused tests for the new message latency metric. This PR adds exported histogram behavior plus clamp/skip semantics, but the current consumer fixtures leave CreatedAt and MsgTimestamp at zero, so e2e/send_to_index emission, zero timestamp skipping, and negative-duration clamping are not covered.

**Comment by lml2468** [comment]
# Code Review — needs-discussion

**Verdict**: `needs-discussion`
**Scope**: 2 files / 30+ / 1-
**Counts**: Must Fix **0** · Should Fix **1** · Nit **1** · Question **0**

---

## Summary

The metric implementation is small and localized: it registers a dedicated histogram, emits only from the successful bulk path, skips zero source timestamps, and clamps negative durations before observation. I did not find a functional correctness blocker in the changed code. The gap is test coverage for the newly exported metric behavior, especially because the existing consumer fixtures leave both timestamp fields at zero.

## Must Fix (0)

无

## Should Fix (1)

- [ ] **[internal/consumer/metrics.go:147]** Add focused tests for `indexer_message_latency_seconds` covering `e2e` and `send_to_index` emission, zero timestamp skipping in the `resolvePass` success path, and negative-duration clamp behavior. — *Impact: Medium, Reach: Moderate*

## Nit (1)

- internal/consumer/metrics.go:20 — The `Metrics` series comment lists the registered consumer series but omits the new `message_latency_seconds{stage}` histogram; update it so the package-level metric inventory stays accurate.

## Question (0)

无

---

## 维度覆盖

| 维度 | 覆盖度 | 主要发现 |
|---|---|---|
| Correctness | full | 0 处 |
| Readability | full | 1 处 nit |
| Maintainability | full | 0 处 |
| Test Coverage | full | 1 处 should fix |

Verified locally: `go test ./internal/consumer/...` and `go test ./...` both pass.

<sub>Labels: `review:done:code:comment` · 由 code-reviewer 智能体生成 · 不代表合并意见，合并是人类专属操作。</sub>

**Comment by lml2468** [comment]
## Review aggregate — **needs-changes**

| reviewer | verdict | comment |
|---|---|---|
| qa-engineer | `fail` | https://github.com/Mininglamp-OSS/octo-search-indexer/pull/41#issuecomment-4832004753 |
| security-engineer | `clear` | https://github.com/Mininglamp-OSS/octo-search-indexer/pull/41#issuecomment-4832035422 |
| code-reviewer | `comment` (needs-discussion) | https://github.com/Mininglamp-OSS/octo-search-indexer/pull/41#issuecomment-4832063926 |

**Aggregate**: `needs-changes` — qa fail (no tests for new metric path) is the highest-priority gate. Code-reviewer flagged the same gap as a Should Fix. Security cleared.

**Action**: applied `needs-human-review` label; this PR requires human triage before merge. Please address the missing tests for `indexer_message_latency_seconds` (e2e/send_to_index emission, zero-timestamp skip, negative-duration clamp), then push and re-request review to restart the loop.
